### PR TITLE
Add entrypoints for moved modules

### PR DESCRIPTION
**Pull request by chrismeyersfsu**

### CODE REVIEW

**Comment by webknjaz** [comment]
I applied formatting in #20 and rebased this change, dropping the pre-commit's conflicting commit.

**Review by webknjaz — COMMENTED** [review]
I imagine that 
https://github.com/ansible/awx-plugins/blob/4126dc7/tests/importable_test.py is coupled tightly enough to warrant being in the scope. Could you also update the smoke test using `@pytest.mark.parametrize`? If you want to keep some preparatory code in the test, add `pytest-subtests` to `dependencies/direct/py.in` and inject the [`subtests`](https://pypi.org/p/pytest-subtests/#subtests-fixture-example) fixture to mark each of the cases in the test report correctly. Though, the same could be achieved by putting that into the own parametrized fixture with `@pytest.fixture(params=[])`. Multiple parametrizations can be stacked. Feel free to ask me if you need pointers.

File: .coveragerc
Comment (chrismeyersfsu) [inline .coveragerc]: Should we reference each module file instead? i.e. `awx_plugins.credentials.aim`

File: tests/importable_test.py
Comment (chrismeyersfsu) [inline tests/importable_test.py]: Should I push these up to a pytest parameterize ?

File: .coveragerc
Comment (webknjaz) [reply .coveragerc]: Maybe. I'm not sure. I haven't tested with multiple modules. But if this works, I'd leave it this way.

I'd even use
```suggestion
  awx_plugins
```
if that's enough to make it function.

Beware that Codecov's v4 action messes up some file paths, though. So don't rely on it, only on the local coverage.py which is integrated through `pytest-cov`.

And with that, I realized I forgot to integrate the HTML report generation into tox :man_facepalming:
Will do that later. But for now, running `tox -- --cov-report=html` should do the trick.

File: dependencies/direct/py.in
Comment (webknjaz) [inline dependencies/direct/py.in]: Are these imported only in the test files directly? If they are runtime deps, move them to the packaging metadata in `pyproject.toml` (`[project].dependencies =`). If some plugins can be optional, their dependencies should probably go into `[project].optional-dependencies =` instead.

Feel free to move the deps updates into a separate PR, though.

File: tests/importable_test.py
Comment (webknjaz) [reply tests/importable_test.py]: Yes. But also, you may want to drop the subprocess stuff. Or maybe split it into a separate test. Tests should assert just one thing usually. It was in my first example but was causing some coverage problems.

Also, look into the newer example where I call `.load()` on the entry point. That's less isolated, but it's exactly what AWX's plugin system would do.

**Review by webknjaz — DISMISSED** [review]
There are a few things needing to be addressed to make this PR mergeable.

**Comment by webknjaz** [comment]
I dropped the automerge until the next review.

File: dependencies/direct/py.in
Comment (chrismeyersfsu) [reply dependencies/direct/py.in]: If I put them in `pyproject.toml` do they ALSO need to be in `py.in` ?

Comment (webknjaz) [reply dependencies/direct/py.in]: `tox` already pulls them in via `package = editable`. And the lock file updater pull in stuff declared in `pyproject.toml`. So everything should be covered by that already.

File: .coveragerc
Comment (webknjaz) [reply .coveragerc]: @chrismeyersfsu I confirmed that this works by looking into the job summary: https://github.com/ansible/awx-plugins/actions/runs/10599668926?pr=17#summary-29375340084.

A little bit of feedback, though: in the future, could you avoid hitting `Resolve` on the threads? Usually, the person starting them is most qualified to judge if they are actually solved. So I sometimes find myself trying to locate such discussions. And when they're hidden, it forces me to click on everything in the UI in hopes of finding the right one.
May I suggest just leaving a comment that it's been addressed? I've seen some folks pasting a commit link or something. If it's really straightforward, and you're sure it's okay to resolve, putting a comment is helpful. This is because clicking `Resolve` doesn't even trigger a notification. I tend to just write “I think X is solved by doing Y. Resolving this thread.” Otherwise, I'd just keep them open until the time comes to merge, at least or until the thread opener closes it.

Anyway, my point is that communication is important for good shared UX/DX.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:85]: Is the inventory plugin missing from here?

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py:26]: Please, don't forget to keep this test. Load all the plugins and verify they're callables.

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py]: I think this triggers `flake8-pytest` violations, but I understand why you missed them. Still, let's try to avoid introducing more.
```suggestion
    (
        'entry_points_group',
        'expected_entry_point_name',
        'expected_entry_point_module',
        'expected_entry_point_func',
    )
```

In general, I expect that any new code, especially files coming from the skeleton, should never introduce new linting violations.

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py]: Technically, the data here is not the `entry_points_group`. It just happens to be the same string, but it's actually coming from the importable.

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py]: This import is unused, it causes a linting rule violation.
```suggestion
```

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py]: same
```suggestion
```

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py]: The last components two are only used in the same place and concatenated always. No need to present them as disconnected inputs, because they are tightly coupled. Plus, taking into account that they are joined to the rest of the import path, I'd say the entire spec should be a param as a single unit.
```suggestion
    ('awx_plugins.credentials', 'aim', 'awx_plugins.credentials.aim:aim_plugin'),
```

Additionally, reusing the `'awx_plugins.credentials'` string this many times will :100: trip the WPS rule violation. It'd need to go into a constant.
Seeing that it's all the same in every param, I'd maybe move it into a separate `@pytest.mark.parametrize`, stacking two decorators together.
OTOH, if the observation of missing inventory plugin references is addressed, they'd need to have this structure. But still, this could be simplified a bit via something like `itertools.product()`.

File: .coveragerc
Comment (webknjaz) [inline .coveragerc]: Feel free to submit this update separately, while figuring out the rest of this PR.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:78]: Should the entry point keys use underscores? I think that since they are not Python identifiers and shouldn't be confused with ones, it's best to use kebab-case. Some projects also use periods in the names.
WDYT?

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:29]: I have a few things to say here:
1) Q: Are all of these mandatory in all installs? Can individual plugins be invoked w/o depending on others? If yes, maybe the deps should be put into a series of extras through `optional-dependencies` rather than unconditional ones.
2) *ACTION REQUIRED:* https://github.com/ansible/awx-plugins/actions/runs/10599668926/job/29375336003?pr=17#step:17:61 revealed that the `awx` dependency is missing. It should be listed somewhere. Perhaps, in an extra, depending on what's decided above.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: tests/importable_test.py
Comment (chrismeyersfsu) [reply tests/importable_test.py]: I never really liked that the first param to pytest.parameterize was a string containing a csv. Good to know that a tuple can be passed.

Comment (webknjaz) [reply tests/importable_test.py]: Either a tuple or a list. I configured the linter to require tuples for consistency. Marking this thread as resolved.

File: pyproject.toml
Comment (chrismeyersfsu) [reply pyproject.toml:78]: my goal is to make as few changes as possible this round. Just get it consumable. Then iterate.

File: .coveragerc
Comment (webknjaz) [reply .coveragerc]: I pulled this bit into `devel`: https://github.com/ansible/awx-plugins/commit/25b3ad16ca681fb60839df3105c41c23a4b6f645. Marking the thread as resolved.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:78]: Okay, consider this postponed. Marking the thread as resolved.

#23

File: tests/importable_test.py
Comment (webknjaz) [inline tests/importable_test.py]: Could you avoid introducing classes? They only complicate things. The tests are already namespaced by being in modules.

If you want to reuse parametrization, you can assign the marker to a variable and use it.
As in
`with_credential_entry_points = pytest.mark.parametrize(...)`
And then, apply `@with_credential_entry_points` to the tests w/o creating another layer of indirection.

**Comment by webknjaz** [comment]
I've decided to suppress some of the import errors in the inventory plugins as they aren't well-integrated anyway. This lets us unblock the docs builds among other things.

The dependencies are added all together and pulled in unconditionally. I think it might make sense to split them into optional ones (extras). I didn't add `awx` as a dependency because it's completely unsuitable for this and opens up a huge can of worms. We'll need to talk about this.

The inventory plugins are exposed as classes while the credential ones are initialized objects. We need to define a consistent interface for all the plugins.

**Pull request merged**
